src/lib/io/sol-memmap-storage-impl-zephyr.c

```diff
@@ -0,0 +1,285 @@
+ * limitations under the License.
+ */
+
+#include <zephyr.h>
+#include <flash.h>
+#include <device.h>
+
+#include "sol-memmap-storage-impl.h"
+
+#include "sol-buffer.h"
+#include "sol-mainloop.h"
+#include "sol-str-slice.h"
+#include "sol-str-table.h"
+#include "sol-util-internal.h"
+#include "sol-vector.h"
+
+static struct device *flash_dev = NULL;
+
+#define ERASE_ALIGN (4096) // see sol_memmap_impl_write_raw()
+#define MAX_RW (256) //every read/write operation is limited to that
+                     //buffer size in Zephyr
+
+int
+sol_memmap_impl_read_raw(struct map_internal *map_internal,
+    const struct sol_memmap_entry *entry,
+    uint64_t mask,
+    struct sol_buffer *buffer)
+{
+    size_t offset, times = 0, extra = 0;
+    uint64_t value = 0;
+    uint32_t i, j;
+    int r;
+
+    offset = FLASH_MEM_REGION_OFFSET + entry->offset;
+
+    /* entry->size may be bigger than the overall useful data on bit
+     * mask cases, but we have to read the whole area anyway */
+    r = sol_buffer_ensure(buffer, entry->size);
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+
+    if (entry->size > MAX_RW) {
+        times = entry->size / MAX_RW;
+        extra = entry->size % MAX_RW;
+    }
+
+    if (!(times || extra)) {
+        r = flash_read(flash_dev, offset, buffer->data, entry->size);
+        SOL_INT_CHECK_GOTO(r, < 0, err);
+    } else {
+        size_t off = 0;
+
+        while (times) {
+            r = flash_read(flash_dev, offset + off, buffer->data + off,
+                MAX_RW);
+            SOL_INT_CHECK_GOTO(r, < 0, err);
+            off += MAX_RW;
+            times--;
+        }
+        if (extra) {
+            r = flash_read(flash_dev, offset + off, buffer->data + off,
+                extra);
+            SOL_INT_CHECK_GOTO(r, < 0, err);
+        }
+    }
+
+    if (mask) {
+        for (i = 0, j = 0; i < entry->size; i++, j += 8)
+            value |= (uint64_t)((uint8_t *)buffer->data)[i] << j;
+
+        value &= mask;
+        value >>= entry->bit_offset;
+
+        memset(buffer->data, 0, buffer->capacity);
+        for (i = 0; i < entry->size; i++, value >>= 8)
+            ((uint8_t *)buffer->data)[i] = value & 0xff;
+    }
+
+    buffer->used = entry->size;
+
+    return 0;
+
+err:
+    sol_buffer_fini(buffer);
+    SOL_WRN("Flash read failed");
+    return r;
+}
+
+/* W25QXXDV memories won't write to any memory region that is not
+ * erased, and the minimum erasable block size is ERASE_ALIGN (also
+ * aligned to that size). We have to read back that much bytes before
+ * any write, and we chose the heap to do so. */
+int
+sol_memmap_impl_write_raw(struct map_internal *map_internal,
+    const char *name,
+    const struct sol_memmap_entry *entry,
+    uint64_t mask,
+    struct sol_blob *blob,
+    void (*cb)(void *data, const char *name, struct sol_blob *blob, int status),
+    const void *data)
+{
+    int r = 0;
+    uint8_t *back_mem;
+    size_t offset, mod, back_sz, cnt;
+
+    if (!sol_blob_ref(blob))
```

Comment:
I guess it's to be safe around the user cb(). I'll make sure we don't leak the ref but still keep it, OK?